src/Integration/Flysystem/S3/src/S3FilesystemV2.php

```diff
@@ -91,31 +93,36 @@ class S3FilesystemV2 implements FilesystemAdapter
      */
     private $visibility;
 
+    /**
+     * @var MimeTypeDetector
+     */
+    private $mimeTypeDetector;
+
     public function __construct(
         S3Client $client,
         string $bucket,
         string $prefix = '',
-        VisibilityConverter $visibility = null
+        VisibilityConverter $visibility = null,
+        MimeTypeDetector $mimeTypeDetector = null
     ) {
         $this->client = $client;
         $this->prefixer = new PathPrefixer($prefix);
         $this->bucket = $bucket;
         $this->visibility = $visibility ?: new PortableVisibilityConverter();
+        $this->mimeTypeDetector = $mimeTypeDetector ?: new FinfoMimeTypeDetector();
     }
 
     public function fileExists(string $path): bool
     {
         try {
-            $this->client->getObject(
+            return $this->client->objectExists(
                 [
                     'Bucket' => $this->bucket,
                     'Key' => $this->prefixer->prefixPath($path),
                 ]
-            )->resolve();
-
-            return true;
+            )->isSuccess();
```

Comment:
This is non-blocking, right?

Reply:
nop, this just returns if the object exist. 
To be blocking (waiting until a final state (either success or failure)) use `$waiter->wait() && $waiter->isSuccess()`

Reply:
Cool cool. 

Then it is all good. It should not be blocking.